// fs-plus/lib/fs-plus.js
(function() {
  var CSON, Module, async, fs, fsExtensions, mkdirp, path, plist, rimraf, statSyncNoException, _,
    __slice = [].slice;

  fs = require('fs');

  Module = require('module');

  path = require('path');

  _ = require('underscore-plus');

  async = require('async');

  CSON = null;

  mkdirp = require('mkdirp');

  plist = null;

  rimraf = require('rimraf');

  fsExtensions = {
    getHomeDirectory: function() {
      if (process.platform === 'win32') {
        return process.env.USERPROFILE;
      } else {
        return process.env.HOME;
      }
    },
    absolute: function(relativePath) {
      var e, homeDir;
      if (relativePath == null) {
        return null;
      }
      homeDir = this.getHomeDirectory();
      if (relativePath === '~') {
        relativePath = homeDir;
      } else if (relativePath.indexOf('~/') === 0) {
        relativePath = "" + homeDir + (relativePath.substring(1));
      }
      try {
        return fs.realpathSync(relativePath);
      } catch (_error) {
        e = _error;
        return relativePath;
      }
    },
    isAbsolute: function(pathToCheck) {
      if (pathToCheck == null) {
        pathToCheck = '';
      }
      if (process.platform === 'win32') {
        return pathToCheck[1] === ':';
      } else {
        return pathToCheck[0] === '/';
      }
    },
    existsSync: function(pathToCheck) {
      return (pathToCheck != null) && statSyncNoException(pathToCheck) !== false;
    },
    isDirectorySync: function(directoryPath) {
      var stat;
      if (!((directoryPath != null ? directoryPath.length : void 0) > 0)) {
        return false;
      }
      if (stat = statSyncNoException(directoryPath)) {
        return stat.isDirectory();
      } else {
        return false;
      }
    },
    isDirectory: function(directoryPath, done) {
      if (!((directoryPath != null ? directoryPath.length : void 0) > 0)) {
        return done(false);
      }
      return fs.exists(directoryPath, function(exists) {
        if (exists) {
          return fs.stat(directoryPath, function(error, stat) {
            if (error != null) {
              return done(false);
            } else {
              return done(stat.isDirectory());
            }
          });
        } else {
          return done(false);
        }
      });
    },
    isFileSync: function(filePath) {
      var stat;
      if (!((filePath != null ? filePath.length : void 0) > 0)) {
        return false;
      }
      if (stat = statSyncNoException(filePath)) {
        return stat.isFile();
      } else {
        return false;
      }
    },
    isSymbolicLinkSync: function(symlinkPath) {
      var error;
      if (!((symlinkPath != null ? symlinkPath.length : void 0) > 0)) {
        return false;
      }
      try {
        return fs.lstatSync(symlinkPath).isSymbolicLink();
      } catch (_error) {
        error = _error;
        return false;
      }
    },
    isExecutableSync: function(pathToCheck) {
      var stat;
      if (!((pathToCheck != null ? pathToCheck.length : void 0) > 0)) {
        return false;
      }
      if (stat = statSyncNoException(pathToCheck)) {
        return (stat.mode & 0x1ff & 1) !== 0;
      } else {
        return false;
      }
    },
    getSizeSync: function(pathToCheck) {
      var _ref;
      if ((pathToCheck != null ? pathToCheck.length : void 0) > 0) {
        return (_ref = statSyncNoException(pathToCheck).size) != null ? _ref : -1;
      } else {
        return -1;
      }
    },
    listSync: function(rootPath, extensions) {
      var paths;
      if (!this.isDirectorySync(rootPath)) {
        return [];
      }
      paths = fs.readdirSync(rootPath);
      if (extensions) {
        paths = this.filterExtensions(paths, extensions);
      }
      paths = paths.map(function(childPath) {
        return path.join(rootPath, childPath);
      });
      return paths;
    },
    list: function() {
      var done, extensions, rest, rootPath,
        _this = this;
      rootPath = arguments[0], rest = 2 <= arguments.length ? __slice.call(arguments, 1) : [];
      if (rest.length > 1) {
        extensions = rest.shift();
      }
      done = rest.shift();
      return fs.readdir(rootPath, function(error, paths) {
        if (error != null) {
          return done(error);
        } else {
          if (extensions) {
            paths = _this.filterExtensions(paths, extensions);
          }
          paths = paths.map(function(childPath) {
            return path.join(rootPath, childPath);
          });
          return done(null, paths);
        }
      });
    },
    filterExtensions: function(paths, extensions) {
      extensions = extensions.map(function(ext) {
        if (ext === '') {
          return ext;
        } else {
          return '.' + ext.replace(/^\./, '');
        }
      });
      return paths.filter(function(pathToCheck) {
        return _.include(extensions, path.extname(pathToCheck));
      });
    },
    listTreeSync: function(rootPath) {
      var onPath, paths;
      paths = [];
      onPath = function(childPath) {
        paths.push(childPath);
        return true;
      };
      this.traverseTreeSync(rootPath, onPath, onPath);
      return paths;
    },
    moveSync: function(source, target) {
      return fs.renameSync(source, target);
    },
    removeSync: function(pathToRemove) {
      return rimraf.sync(pathToRemove);
    },
    writeFileSync: function(filePath, content, options) {
      mkdirp.sync(path.dirname(filePath));
      return fs.writeFileSync(filePath, content, options);
    },
    writeFile: function(filePath, content, options, callback) {
      callback = _.last(arguments);
      return mkdirp(path.dirname(filePath), function(error) {
        if (error != null) {
          return typeof callback === "function" ? callback(error) : void 0;
        } else {
          return fs.writeFile(filePath, content, options, callback);
        }
      });
    },
    copy: function(sourcePath, destinationPath, done) {
      return mkdirp(path.dirname(destinationPath), function(error) {
        var destinationStream, sourceStream;
        if (error != null) {
          if (typeof done === "function") {
            done(error);
          }
          return;
        }
        sourceStream = fs.createReadStream(sourcePath);
        sourceStream.on('error', function(error) {
          if (typeof done === "function") {
            done(error);
          }
          return done = null;
        });
        destinationStream = fs.createWriteStream(destinationPath);
        destinationStream.on('error', function(error) {
          if (typeof done === "function") {
            done(error);
          }
          return done = null;
        });
        destinationStream.on('close', function() {
          if (typeof done === "function") {
            done();
          }
          return done = null;
        });
        return sourceStream.pipe(destinationStream);
      });
    },
    copySync: function(sourcePath, destinationPath) {
      var content, destinationFilePath, source, sourceFilePath, _i, _len, _ref, _results;
      mkdirp.sync(destinationPath);
      _ref = fs.readdirSync(sourcePath);
      _results = [];
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        source = _ref[_i];
        sourceFilePath = path.join(sourcePath, source);
        destinationFilePath = path.join(destinationPath, source);
        if (this.isDirectorySync(sourceFilePath)) {
          _results.push(this.copySync(sourceFilePath, destinationFilePath));
        } else {
          content = fs.readFileSync(sourceFilePath);
          _results.push(fs.writeFileSync(destinationFilePath, content));
        }
      }
      return _results;
    },
    makeTreeSync: function(directoryPath) {
      if (directoryPath && !this.existsSync(directoryPath)) {
        return mkdirp.sync(directoryPath);
      }
    },
    traverseTreeSync: function(rootPath, onFile, onDirectory) {
      var traverse;
      if (onDirectory == null) {
        onDirectory = onFile;
      }
      if (!this.isDirectorySync(rootPath)) {
        return;
      }
      traverse = function(directoryPath, onFile, onDirectory) {
        var childPath, file, linkStats, stats, _i, _len, _ref, _results;
        _ref = fs.readdirSync(directoryPath);
        _results = [];
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          file = _ref[_i];
          childPath = path.join(directoryPath, file);
          stats = fs.lstatSync(childPath);
          if (stats.isSymbolicLink()) {
            if (linkStats = statSyncNoException(childPath)) {
              stats = linkStats;
            }
          }
          if (stats.isDirectory()) {
            if (onDirectory(childPath)) {
              _results.push(traverse(childPath, onFile, onDirectory));
            } else {
              _results.push(void 0);
            }
          } else if (stats.isFile()) {
            _results.push(onFile(childPath));
          } else {
            _results.push(void 0);
          }
        }
        return _results;
      };
      return traverse(rootPath, onFile, onDirectory);
    },
    traverseTree: function(rootPath, onFile, onDirectory, onDone) {
      return fs.readdir(rootPath, function(error, files) {
        var file, queue, _i, _len, _results;
        if (error) {
          return typeof onDone === "function" ? onDone() : void 0;
        } else {
          queue = async.queue(function(childPath, callback) {
            return fs.stat(childPath, function(error, stats) {
              if (error) {
                return callback(error);
              } else if (stats.isFile()) {
                onFile(childPath);
                return callback();
              } else if (stats.isDirectory()) {
                if (onDirectory(childPath)) {
                  return fs.readdir(childPath, function(error, files) {
                    var file, _i, _len;
                    if (error) {
                      return callback(error);
                    } else {
                      for (_i = 0, _len = files.length; _i < _len; _i++) {
                        file = files[_i];
                        queue.unshift(path.join(childPath, file));
                      }
                      return callback();
                    }
                  });
                } else {
                  return callback();
                }
              }
            });
          });
          queue.concurrency = 1;
          queue.drain = onDone;
          _results = [];
          for (_i = 0, _len = files.length; _i < _len; _i++) {
            file = files[_i];
            _results.push(queue.push(path.join(rootPath, file)));
          }
          return _results;
        }
      });
    },
    md5ForPath: function(pathToDigest) {
      var contents;
      contents = fs.readFileSync(pathToDigest);
      return require('crypto').createHash('md5').update(contents).digest('hex');
    },
    resolve: function() {
      var args, candidatePath, extensions, loadPath, loadPaths, pathToResolve, resolvedPath, _i, _len;
      args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
      if (_.isArray(_.last(args))) {
        extensions = args.pop();
      }
      pathToResolve = args.pop();
      loadPaths = args;
      if (this.isAbsolute(pathToResolve)) {
        if (extensions && (resolvedPath = this.resolveExtension(pathToResolve, extensions))) {
          return resolvedPath;
        } else {
          if (this.existsSync(pathToResolve)) {
            return pathToResolve;
          }
        }
      }
      for (_i = 0, _len = loadPaths.length; _i < _len; _i++) {
        loadPath = loadPaths[_i];
        candidatePath = path.join(loadPath, pathToResolve);
        if (extensions) {
          if (resolvedPath = this.resolveExtension(candidatePath, extensions)) {
            return resolvedPath;
          }
        } else {
          if (this.existsSync(candidatePath)) {
            return this.absolute(candidatePath);
          }
        }
      }
      return void 0;
    },
    resolveOnLoadPath: function() {
      var args, loadPaths;
      args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
      loadPaths = Module.globalPaths.concat(module.paths);
      return this.resolve.apply(this, __slice.call(loadPaths).concat(__slice.call(args)));
    },
    resolveExtension: function(pathToResolve, extensions) {
      var extension, pathWithExtension, _i, _len;
      for (_i = 0, _len = extensions.length; _i < _len; _i++) {
        extension = extensions[_i];
        if (extension === "") {
          if (this.existsSync(pathToResolve)) {
            return this.absolute(pathToResolve);
          }
        } else {
          pathWithExtension = pathToResolve + "." + extension.replace(/^\./, "");
          if (this.existsSync(pathWithExtension)) {
            return this.absolute(pathWithExtension);
          }
        }
      }
      return void 0;
    },
    isCompressedExtension: function(ext) {
      return _.indexOf(['.gz', '.jar', '.tar', '.tgz', '.zip'], ext, true) >= 0;
    },
    isImageExtension: function(ext) {
      return _.indexOf(['.gif', '.jpeg', '.jpg', '.png', '.tiff'], ext, true) >= 0;
    },
    isPdfExtension: function(ext) {
      return ext === '.pdf';
    },
    isBinaryExtension: function(ext) {
      return _.indexOf(['.DS_Store', '.a', '.o', '.so', '.woff'], ext, true) >= 0;
    },
    isReadmePath: function(readmePath) {
      var base, extension;
      extension = path.extname(readmePath);
      base = path.basename(readmePath, extension).toLowerCase();
      return base === 'readme' && (extension === '' || this.isMarkdownExtension(extension));
    },
    isMarkdownExtension: function(ext) {
      return _.indexOf(['.markdown', '.md', '.mdown', '.mkd', '.mkdown', '.ron'], ext, true) >= 0;
    },
    readObjectSync: function(objectPath) {
      if (CSON == null) {
        CSON = require('season');
      }
      if (CSON.isObjectPath(objectPath)) {
        return CSON.readFileSync(objectPath);
      } else {
        return this.readPlistSync(objectPath);
      }
    },
    readObject: function(objectPath, done) {
      if (CSON == null) {
        CSON = require('season');
      }
      if (CSON.isObjectPath(objectPath)) {
        return CSON.readFile(objectPath, done);
      } else {
        return this.readPlist(objectPath, done);
      }
    },
    readPlistSync: function(plistPath) {
      if (plist == null) {
        plist = require('plist');
      }
      return plist.parseStringSync(fs.readFileSync(plistPath, 'utf8'));
    },
    readPlist: function(plistPath, done) {
      if (plist == null) {
        plist = require('plist');
      }
      return fs.readFile(plistPath, 'utf8', function(error, contents) {
        var parseError;
        if (error != null) {
          return done(error);
        } else {
          try {
            return done(null, plist.parseStringSync(contents));
          } catch (_error) {
            parseError = _error;
            return done(parseError);
          }
        }
      });
    }
  };

  statSyncNoException = fs.statSyncNoException;

  if (statSyncNoException == null) {
    statSyncNoException = function() {
      var args, error;
      args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
      try {
        return fs.statSync.apply(fs, args);
      } catch (_error) {
        error = _error;
        return false;
      }
    };
  }

  module.exports = _.extend({}, fs, fsExtensions);

}).call(this);
